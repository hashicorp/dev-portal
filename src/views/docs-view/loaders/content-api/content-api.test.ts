/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import nock from 'nock'

import * as contentApi from '.'

import document200 from './__fixtures__/document_200.json'
import document404 from './__fixtures__/document_404.json'
import navData200 from './__fixtures__/navData_200.json'
import navData404 from './__fixtures__/navData_404.json'
import versionMetadata200 from './__fixtures__/versionMetadata_200.json'
import versionMetadataEmpty from './__fixtures__/versionMetadata_empty.json'

describe('contentApi', () => {
	let scope: nock.Scope

	beforeEach(() => {
		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		scope = nock(process.env.MKTG_CONTENT_API!)
	})

	describe('fetchDocument', () => {
		it('should return the requested document', async () => {
			scope
				.get('/api/content/waypoint/doc/v0.5.x/commands/k8s-bootstrap')
				.reply(200, document200)

			const res = await contentApi.fetchDocument(
				'waypoint',
				'doc/v0.5.x/commands/k8s-bootstrap'
			)
			expect(res).toMatchSnapshot()
		})

		it('should throw if a document is not found', async () => {
			scope
				.get('/api/content/waypoint/doc/latest/commands/k8s-bootstrap')
				.reply(404, document404)

			await expect(async () => {
				await contentApi.fetchDocument(
					'waypoint',
					'doc/latest/commands/k8s-bootstrap'
				)
			}).rejects.toThrowErrorMatchingSnapshot()
		})
	})

	describe('fetchNavData', () => {
		it('should return the requested navdata', async () => {
			scope
				.get('/api/content/waypoint/nav-data/v0.5.x/commands')
				.reply(200, navData200)

			const res = await contentApi.fetchNavData(
				'waypoint',
				'commands',
				'v0.5.x'
			)
			expect(res).toMatchSnapshot()
		})

		it('should throw if a navData is not found', async () => {
			scope
				.get('/api/content/waypoint/nav-data/v0.9000.x/commands')
				.reply(404, navData404)

			await expect(async () => {
				await contentApi.fetchNavData('waypoint', 'commands', 'v0.9000.x')
			}).rejects.toThrowErrorMatchingSnapshot()
		})
	})

	describe('fetchVersionMetadataList', () => {
		it('should return the requested version metadata list', async () => {
			scope
				.get('/api/content/waypoint/version-metadata')
				.query({ partial: true })
				.reply(200, versionMetadata200)

			const res = await contentApi.fetchVersionMetadataList('waypoint')
			expect(res).toMatchSnapshot()
		})

		it('should return an empty list for invalid product name', async () => {
			scope
				.get('/api/content/waypointzzz/version-metadata')
				.query({ partial: true })
				.reply(200, versionMetadataEmpty)

			const res = await contentApi.fetchVersionMetadataList('waypointzzz')
			expect(res).toMatchSnapshot()
		})
	})
})
