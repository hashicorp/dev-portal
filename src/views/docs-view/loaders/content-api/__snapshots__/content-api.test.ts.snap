// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`contentApi > fetchDocument > should return the requested document 1`] = `
{
  "created_at": "Sat Oct 02 2021 23:52:13 GMT+0000 (Coordinated Universal Time)",
  "fullPath": "doc/v0.5.x/commands",
  "githubFile": "website/content/commands/index.mdx",
  "markdownSource": "# Waypoint Commands (CLI)

Waypoint is controlled via a very easy to use command-line interface (CLI). Waypoint
is only a single command-line application: \`waypoint\`. This application then takes a
subcommand such as \`artifact\` or \`deployment\`. The complete list of subcommands is in
the navigation to the left.

The waypoint CLI is a well-behaved command line application. In erroneous cases, a
non-zero exit status will be returned. It also responds to -h and --help as you'd most
likely expect.

To view a list of the available commands at any time, just run waypoint with no arguments:

\`\`\`shell-session
$ waypoint
Usage: waypoint [-version] [-help] [-autocomplete-(un)install] <command> [args]

Common commands:
    build        Build a new versioned artifact from source
    deploy       Deploy a pushed artifact
    release      Release a deployment
    up           Perform the build, deploy, and release steps for the app

Other commands:
    artifact        Artifact and build management
    auth-method     Auth method management
    config          Application configuration management
    context         Server access configurations
    deployment      Deployment creation and management
    destroy         Delete all the resources created for an app
    docs            Show documentation for components
    exec            Execute a command in the context of a running application instance
    fmt             Rewrite waypoint.hcl configuration to a canonical format
    hostname        Application URLs
    init            Initialize and validate a project
    install         Install the Waypoint server to Kubernetes, Nomad, or Docker
    logs            Show log output from the current application deployment
    project         Project management
    runner          Runner management
    server          Server management
    ui              Open the web UI
    user            User information and management
    version         Prints the version of this Waypoint CLI
\`\`\`

To get help for any specific command, pass the -h flag to the relevant subcommand.
For example, to see help about the up subcommand:

\`\`\`shell-session
$ waypoint up -h
Usage: waypoint up [options]

  Perform the build, deploy, and release steps for the app.

Global Options:

  -workspace=<string>
      Workspace to operate in. Defaults to 'default'. The default is default.

Operation Options:

  -label=<key=value>
      Labels to set for this operation. Can be specified multiple times.

  -remote
      True to use a remote runner to execute. This defaults to false unless
      'runner.default' is set in your configuration. The default is false.

  -remote-source=<key=value>
      Override configurations for how remote runners source data. This is
      specified to the data source type being used in your configuration. This
      is used for example to set a specific Git ref to run against.
\`\`\`

Running \`waypoint -autocomplete-install\` will add the waypoint autocomplete capability
so you can do \`waypoint <tab>\` on commands. Running \`waypoint -autocomplete-uninstall\`
will remove it. Please note that this will modify your shell init script, so you will
need to reload your shell.
",
  "mdxTransforms": [
    "extract-frontmatter",
    "include-partials",
    "paragraph-custom-alerts",
  ],
  "metadata": {
    "layout": "commands",
    "page_title": "Waypoint Commands (CLI)",
  },
  "product": "waypoint",
  "sha": "f441834f6a78b94c9108f1bf348c793b3bba08a6",
  "subpath": "commands",
  "version": "v0.5.x",
}
`;

exports[`contentApi > fetchDocument > should throw if a document is not found 1`] = `[ContentApiError: Failed to fetch: https://content.hashicorp.com/api/content/waypoint/doc/latest/commands/k8s-bootstrap]`;

exports[`contentApi > fetchNavData > should return the requested navdata 1`] = `
{
  "created_at": "Sat Oct 02 2021 23:52:11 GMT+0000 (Coordinated Universal Time)",
  "fullPath": "nav-data/v0.5.x/commands",
  "githubFile": "website/content/commands-nav-data.json",
  "navData": [
    {
      "path": "build",
      "title": "build",
    },
    {
      "path": "deploy",
      "title": "deploy",
    },
    {
      "path": "context",
      "title": "context",
    },
    {
      "path": "context-inspect",
      "title": "context-inspect",
    },
    {
      "path": "destroy",
      "title": "destroy",
    },
    {
      "path": "exec",
      "title": "exec",
    },
    {
      "path": "init",
      "title": "init",
    },
    {
      "path": "install",
      "title": "install",
    },
    {
      "path": "login",
      "title": "login",
    },
    {
      "path": "logs",
      "title": "logs",
    },
    {
      "path": "project-apply",
      "title": "project apply",
    },
    {
      "path": "release",
      "title": "release",
    },
    {
      "path": "ui",
      "title": "ui",
    },
    {
      "path": "up",
      "title": "up",
    },
    {
      "divider": true,
    },
    {
      "path": "artifact-build",
      "title": "artifact build",
    },
    {
      "path": "artifact-list-builds",
      "title": "artifact list-builds",
    },
    {
      "path": "artifact-list",
      "title": "artifact list",
    },
    {
      "path": "artifact-push",
      "title": "artifact push",
    },
    {
      "path": "auth-method-list",
      "title": "auth-method list",
    },
    {
      "path": "auth-method-inspect",
      "title": "auth-method inspect",
    },
    {
      "path": "auth-method-set-oidc",
      "title": "auth-method set oidc",
    },
    {
      "path": "auth-method-delete",
      "title": "auth-method delete",
    },
    {
      "path": "config-get",
      "title": "config get",
    },
    {
      "path": "config-set",
      "title": "config set",
    },
    {
      "path": "config-source-get",
      "title": "config source-get",
    },
    {
      "path": "config-source-set",
      "title": "config source-set",
    },
    {
      "path": "config-sync",
      "title": "config sync",
    },
    {
      "path": "context-clear",
      "title": "context clear",
    },
    {
      "path": "context-create",
      "title": "context create",
    },
    {
      "path": "context-delete",
      "title": "context delete",
    },
    {
      "path": "context-list",
      "title": "context list",
    },
    {
      "path": "context-rename",
      "title": "context rename",
    },
    {
      "path": "context-use",
      "title": "context use",
    },
    {
      "path": "context-verify",
      "title": "context verify",
    },
    {
      "path": "deployment-deploy",
      "title": "deployment deploy",
    },
    {
      "path": "deployment-destroy",
      "title": "deployment destroy",
    },
    {
      "path": "deployment-list",
      "title": "deployment list",
    },
    {
      "path": "docs",
      "title": "docs",
    },
    {
      "path": "fmt",
      "title": "fmt",
    },
    {
      "path": "hostname-delete",
      "title": "hostname delete",
    },
    {
      "path": "hostname-list",
      "title": "hostname list",
    },
    {
      "path": "hostname-register",
      "title": "hostname register",
    },
    {
      "path": "plugin",
      "title": "plugin",
    },
    {
      "path": "project-inspect",
      "title": "project inspect",
    },
    {
      "path": "project-list",
      "title": "project list",
    },
    {
      "path": "release-list",
      "title": "release list",
    },
    {
      "path": "runner-agent",
      "title": "runner agent",
    },
    {
      "path": "server-bootstrap",
      "title": "server bootstrap",
    },
    {
      "path": "server-config-set",
      "title": "server config-set",
    },
    {
      "path": "server-install",
      "title": "server install",
    },
    {
      "path": "server-restore",
      "title": "server restore",
    },
    {
      "path": "server-run",
      "title": "server run",
    },
    {
      "path": "server-snapshot",
      "title": "server snapshot",
    },
    {
      "path": "server-uninstall",
      "title": "server uninstall",
    },
    {
      "path": "server-upgrade",
      "title": "server upgrade",
    },
    {
      "path": "status",
      "title": "status",
    },
    {
      "path": "token-exchange",
      "title": "token exchange",
    },
    {
      "path": "token-invite",
      "title": "token invite",
    },
    {
      "path": "token-new",
      "title": "token new",
    },
    {
      "path": "user-inspect",
      "title": "user inspect",
    },
    {
      "path": "user-invite",
      "title": "user invite",
    },
    {
      "path": "user-modify",
      "title": "user modify",
    },
    {
      "path": "user-token",
      "title": "user token",
    },
    {
      "path": "version",
      "title": "version",
    },
  ],
  "product": "waypoint",
  "sha": "f441834f6a78b94c9108f1bf348c793b3bba08a6",
  "subpath": "commands",
  "version": "v0.5.x",
}
`;

exports[`contentApi > fetchNavData > should throw if a navData is not found 1`] = `[ContentApiError: Failed to fetch: https://content.hashicorp.com/api/content/waypoint/nav-data/v0.9000.x/commands]`;

exports[`contentApi > fetchVersionMetadataList > should return an empty list for invalid product name 1`] = `[]`;

exports[`contentApi > fetchVersionMetadataList > should return the requested version metadata list 1`] = `
[
  {
    "created_at": "Fri Sep 24 2021 07:56:45 GMT+0000 (Coordinated Universal Time)",
    "display": "v0.6.2",
    "fullPath": "version-metadata/v0.6.x",
    "isLatest": false,
    "product": "waypoint",
    "ref": "kevin/versioned-docs",
    "releaseStage": "alpha",
    "sha": "e14900057317f9fd63eae31f3414a0ba383faa8e",
    "version": "v0.6.x",
  },
  {
    "created_at": "Thu Sep 23 2021 07:56:45 GMT+0000 (Coordinated Universal Time)",
    "display": "v0.5.2",
    "fullPath": "version-metadata/v0.5.x",
    "isLatest": true,
    "product": "waypoint",
    "ref": "kevin/versioned-docs",
    "releaseStage": "stable",
    "sha": "214900057317f9fd63eae31f3414a0ba383faa9e",
    "version": "v0.5.x",
  },
  {
    "created_at": "Tue Sep 21 2021 21:05:43 GMT+0000 (Coordinated Universal Time)",
    "fullPath": "version-metadata/v0.4.x",
    "product": "waypoint",
    "ref": "release/0.4.x",
    "releaseStage": "stable",
    "sha": "c6cd81e10f30d339c113772b5f978d2f96cd1a2e",
    "version": "v0.4.x",
  },
  {
    "created_at": "Wed Sep 22 2021 17:40:36 GMT+0000 (Coordinated Universal Time)",
    "fullPath": "version-metadata/v0.3.x",
    "product": "waypoint",
    "ref": "refs/heads/release/0.3.x",
    "releaseStage": "stable",
    "sha": "61a1d140146321a8d03859ba8b336bd5baa55f67",
    "version": "v0.3.x",
  },
]
`;
